File: Backend/controllers/reportController.js
const Data = require('../models/dataModel');
const Prediction = require('../models/predictionModel');
const History = require('../models/predictionInputModel');
const { calculateRiskScore } = require('../utils/helpers');
const XLSX = require('xlsx');

const ReportController = {
  getFullReport: async (req, res) => {
    try {
      const data = await Data.findAll(10000, 0, req.user.id);
      const predictions = await Prediction.findAll(req.user.id);
      const history = await History.findAll(req.user.id);
      const stats = await Data.getStats(req.user.id);
      const predStats = await Prediction.getStats(req.user.id);
      const histStats = await History.getStats(req.user.id);

      const enriched = data.map(d => {
        const risk = calculateRiskScore(d.suhu, d.pH, d.salinitas, d.kekeruhan);
        return { ...d, skor_risiko: risk.skor, status: risk.status };
      });

      res.json({
        success: true,
        data: {
          summary: { ...stats, ...histStats, ...predStats },
          data: enriched,
          predictions,
          history
        }
      });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },

  exportExcel: async (req, res) => {
    try {
      const { lokasi = '' } = req.query;
      const includePrediksi = req.query.prediksi !== 'false';

      const wb = XLSX.utils.book_new();

      if (includePrediksi) {
        const predictions = await Prediction.findAll(req.user.id);
        const filtered = lokasi ? predictions.filter((p) => p.lokasi === lokasi) : predictions;
        const sorted = [...filtered].sort((a, b) => (a.lokasi || '').localeCompare(b.lokasi || '') || Number(a.step_ke) - Number(b.step_ke));
        const rows = sorted.map((p) => ({
          'Tanggal Prediksi': p.tanggal_prediksi,
          Lokasi: p.lokasi,
          'Step Ke': p.step_ke,
          Suhu: p.suhu,
          pH: p.pH,
          Salinitas: p.salinitas,
          Kekeruhan: p.kekeruhan,
          'Skor Risiko': p.skor_risiko,
          Status: p.status,
        }));
        XLSX.utils.book_append_sheet(wb, XLSX.utils.json_to_sheet(rows), 'Prediksi');
      }

      const data = await Data.findAllExport(req.user.id, lokasi || null);
      const dataRows = data.map((d) => ({
        Tanggal: d.tanggal,
        Lokasi: d.lokasi,
        Suhu: d.suhu,
        pH: d.pH,
        Salinitas: d.salinitas,
        Kekeruhan: d.kekeruhan,
      }));
      XLSX.utils.book_append_sheet(
        wb,
        XLSX.utils.json_to_sheet(dataRows.length ? dataRows : [{ Tanggal: 'Tidak ada data kualitas air' }]),
        'Data Kualitas Air'
      );

      const buffer = XLSX.write(wb, { type: 'buffer', bookType: 'xlsx' });
      res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
      res.setHeader('Content-Disposition', `attachment; filename=laporan_${Date.now()}.xlsx`);
      res.send(buffer);
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  }
};

module.exports = ReportController;

File: backend_python/controllers/report_controller.py
import time
from io import BytesIO

from flask import g, request
from openpyxl import Workbook

from models.data_model import Data
from models.prediction_input_model import PredictionInput
from models.prediction_model import Prediction
from utils.helpers import calculate_risk_score, fail, jsonable, ok


class ReportController:
    @staticmethod
    def get_full_report():
        rows = Data.find_all(10000, 0, g.user["id"])
        predictions = Prediction.find_all(g.user["id"])
        history = PredictionInput.find_all(g.user["id"])
        stats = Data.get_stats(g.user["id"])
        pred_stats = Prediction.get_stats(g.user["id"])
        hist_stats = PredictionInput.get_stats(g.user["id"])

        enriched = []
        for d in rows:
            risk = calculate_risk_score(d["suhu"], d["pH"], d["salinitas"], d["kekeruhan"])
            enriched.append({**d, "skor_risiko": risk["skor"], "status": risk["status"]})

        summary = {**(stats or {}), **(hist_stats or {}), **(pred_stats or {})}
        payload = {
            "success": True,
            "data": {
                "summary": summary,
                "data": enriched,
                "predictions": predictions,
                "history": history,
            },
        }
        from flask import jsonify

        return jsonify(jsonable(payload))

    @staticmethod
    def export_excel():
        lokasi = request.args.get("lokasi", "")
        include_prediksi = request.args.get("prediksi") != "false"

        wb = Workbook()

        if include_prediksi:
            predictions = Prediction.find_all(g.user["id"])
            filtered = [p for p in predictions if p.get("lokasi") == lokasi] if lokasi else predictions
            sorted_rows = sorted(
                filtered,
                key=lambda p: ((p.get("lokasi") or ""), int(p["step_ke"])),
            )
            ws = wb.active
            ws.title = "Prediksi"
            headers = ["Tanggal Prediksi", "Lokasi", "Step Ke", "Suhu", "pH", "Salinitas", "Kekeruhan", "Skor Risiko", "Status"]
            ws.append(headers)
            for p in sorted_rows:
                ws.append(
                    [
                        str(p["tanggal_prediksi"]),
                        p["lokasi"],
                        p["step_ke"],
                        float(p["suhu"]),
                        float(p["pH"]),
                        float(p["salinitas"]),
                        float(p["kekeruhan"]),
                        float(p["skor_risiko"]),
                        p["status"],
                    ]
                )

        data = Data.find_all_export(g.user["id"], lokasi or None)
        ws2 = wb.create_sheet("Data Kualitas Air")
        ws2.append(["Tanggal", "Lokasi", "Suhu", "pH", "Salinitas", "Kekeruhan"])
        if data:
            for d in data:
                tanggal = d["tanggal"]
                ws2.append(
                    [
                        str(tanggal),
                        d["lokasi"],
                        float(d["suhu"]),
                        float(d["pH"]),
                        float(d["salinitas"]),
                        float(d["kekeruhan"]),
                    ]
                )
        else:
            ws2.append(["Tidak ada data kualitas air"])

        buffer = BytesIO()
        wb.save(buffer)
        body = buffer.getvalue()
        filename = f"laporan_{int(time.time() * 1000)}.xlsx"

        from flask import Response

        response = Response(body)
        response.headers["Content-Type"] = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
        response.headers["Content-Disposition"] = f"attachment; filename={filename}"
        return response
